### PR TITLE
Spans now derive equality

### PR DESCRIPTION
**Pull request by DanielJCampbell**
r? @nrc

### CODE REVIEW

**Comment by petrochenkov** [comment]
Why is this needed if derived equality is never used (but now can be used accidentally)?

**Comment by DanielJCampbell** [comment]
Spans derive hashing, and are used in HashSets - I felt it strange to derive hashing and not derive equality.

File: src/libsyntax/codemap.rs [hunk 1]
```diff
@@ -151,13 +151,17 @@ pub const COMMAND_LINE_SP: Span = Span { lo: BytePos(0),
...
-        if self == DUMMY_SP { other } else { self }
+        if self.source_equal(&DUMMY_SP) { other } else { self }
     }
 
     pub fn contains(self, other: Span) -> bool {
         self.lo <= other.lo && other.hi <= self.hi
     }
 
+    pub fn source_equal(&self, other: &Span) -> bool {
+        (*self).lo == (*other).lo && (*self).hi == (*other).hi
```
Comment (oli-obk) [inline src/libsyntax/codemap.rs]: The dereferencing is unnecessary. Just let autoderef do its job

**Comment by petrochenkov** [comment]
>Spans derive hashing, and are used in HashSets ... and not derive equality.

I can't check right now, but this usually means that hashes are incorrect and the hash sets work by chance.

**Comment by mitaa** [comment]
This comment/code should probably be updated as well
https://github.com/DanielJCampbell/rust/blob/SpanEquality/src/libsyntax/errors/emitter.rs#L178-L180

**Comment by DanielJCampbell** [comment]
Good catch, I'll update that in the morning.

**Comment by nrc** [comment]
@petrochenkov yes, without this PR hashing for spans is probably buggy. The point of this PR is to fix that so that `==` for spans is a structural equality test.

File: src/libsyntax/codemap.rs [hunk 2]
```diff
@@ -151,13 +151,17 @@ pub const COMMAND_LINE_SP: Span = Span { lo: BytePos(0),
...
     pub fn substitute_dummy(self, other: Span) -> Span {
-        if self == DUMMY_SP { other } else { self }
+        if self.source_equal(&DUMMY_SP) { other } else { self }
     }
 
     pub fn contains(self, other: Span) -> bool {
         self.lo <= other.lo && other.hi <= self.hi
     }
 
+    pub fn source_equal(&self, other: &Span) -> bool {
```
Comment (nrc) [inline src/libsyntax/codemap.rs]: Could you add a comment stating when to use this vs when to use `==`

**Comment by DanielJCampbell** [comment]
Suggested changes made

**Pull request merged**